src/IO/WriteBufferFromPocoSocketChunked.cpp

```diff
@@ -0,0 +1,207 @@
+#include <IO/WriteBufferFromPocoSocketChunked.h>
+#include <Common/logger_useful.h>
+#include <IO/NetUtils.h>
+
+
+namespace
+{
+
+template <typename T>
+const T & setValue(T * typed_ptr, std::type_identity_t<T> val)
+{
+    memcpy(typed_ptr, &val, sizeof(T));
+    return *typed_ptr;
+}
+
+}
+
+namespace DB
+{
+
+void WriteBufferFromPocoSocketChunked::enableChunked()
+{
+    chunked = true;
+    /// Initialize next chunk
+    chunk_size_ptr = reinterpret_cast<decltype(chunk_size_ptr)>(pos);
+    pos += std::min(available(), sizeof(*chunk_size_ptr));
+}
+
+void WriteBufferFromPocoSocketChunked::finishChunk()
+{
+    if (!chunked)
+        return;
+
+    if (pos <= reinterpret_cast<Position>(chunk_size_ptr) + sizeof(*chunk_size_ptr))
+    {
+        if (chunk_size_ptr == last_finish_chunk) // prevent duplicate finish chunk
+            return;
+
+        /// If current chunk is empty it means we are finishing a chunk previously sent by next(),
+        /// we want to convert current chunk header into end-of-chunk marker and initialize next chunk.
+        /// We don't need to worry about if it's the end of the buffer because next() always sends the whole buffer
+        /// so it should be a beginning of the buffer.
+
+        chassert(reinterpret_cast<Position>(chunk_size_ptr) == working_buffer.begin());
+
+        setValue(chunk_size_ptr, 0);
+        /// Initialize next chunk
+        chunk_size_ptr = reinterpret_cast<decltype(chunk_size_ptr)>(pos);
+        pos += std::min(available(), sizeof(*chunk_size_ptr));
+
+        last_finish_chunk = chunk_size_ptr;
+
+        return;
+    }
+
+    /// Previously finished chunk wasn't sent yet
+    if (last_finish_chunk == chunk_size_ptr)
+    {
+        chunk_started = false;
+        LOG_TEST(log, "{} -> {} Chunk send ended.", ourAddress().toString(), peerAddress().toString());
+    }
+
+    /// Fill up current chunk size
+    setValue(chunk_size_ptr, toLittleEndian(static_cast<UInt32>(pos - reinterpret_cast<Position>(chunk_size_ptr) - sizeof(*chunk_size_ptr))));
+
+    if (!chunk_started)
+        LOG_TEST(log, "{} -> {} Chunk send started. Message {}, size {}",
+                ourAddress().toString(), peerAddress().toString(),
+                static_cast<unsigned int>(*(reinterpret_cast<char *>(chunk_size_ptr) + sizeof(*chunk_size_ptr))),
+                *chunk_size_ptr);
+    else
+    {
+        chunk_started = false;
+        LOG_TEST(log, "{} -> {} Chunk send continued. Size {}", ourAddress().toString(), peerAddress().toString(), *chunk_size_ptr);
+    }
+
+    LOG_TEST(log, "{} -> {} Chunk send ended.", ourAddress().toString(), peerAddress().toString());
+
+    if (available() < sizeof(*chunk_size_ptr))
+    {
+        finishing = available();
+        pos += available();
+        chunk_size_ptr = reinterpret_cast<decltype(chunk_size_ptr)>(pos);
```

Comment:
Why do we set `chunk_size_ptr` here? It seems to be unnecessary, because in nextImpl we will reset it when `finishing < sizeof(*chunk_size_ptr)`

Reply:
it is required for detection of second consequent call of `finishChunk`...
besides, I prefer to keep state always valid despite if it needed or not :)

Reply:
yes, you are right, it's a bug